SCP 2345: Replace ContractExe with Builtin contracts in plutus-pab

koslambrou opened the pull request:
Replaced `ContractExe` with `Builtin a` in `plutus-pab`. Here's a resume of what's been done:

- Deleted `Plutus.PAB.ContractCLI` and `Plutus.PAB.Effects.Contract.ContractExe`
- Replaced all occurrences of `ContractExe` with `Builtin a`
- Changed the executable name of `plutus-pab` to `plutus-pab-setup` and only provides the `PSGenerator` command, the migrate command and the `defaultConfig` command.
- Added the module `Plutus.PAB.Run` and `Plutus.PAB.Run.hs` which are used to make an `IO ()` action specialized to a specific `Builtin`. This `IO` action basically replaces the `Plutus.PAB.ContractCLI` module.
- Created the executable `plutus-pab-examples` which provides the available example contracts as `Builtin`.
- Removed the old executables in `plutus-pab` which use `Plutus.PAB.ContractCLI`
- Changed the executable `marlowe-pab` to run the `Builtin` contract `Marlowe`, `MarloweFollow` and `WalletCompanion`. Removed the old Marlowe executables using `Plutus.PAB.ContractCLI`.
- Separated the `PSGenerator` in two parts. The first generates general PS modules, and the second generates PS modules related to a specific `Builtin`. The first can be run with `plutus-pab-setup`, but the other can only run with a specific `Builtin`.
- Removed the possibility to store contract executables in `plutus-pab`.
- Modified nix files to remove used of installed contracts.
- Each specific `Builtin` contracts now have to specify their definitions via `HasDefinitions`.
- Each specific `Builtin` contracts now have to specify the list of modules that need to be converted to PS via `HasPSTypes`.

<!--
Here are some checklists you may like to use. Use your judgement.

This is just a checklist, all the normative suggestions are covered in more detail in CONTRIBUTING.
-->
Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

j-mueller commented on plutus-pab/examples/uniswap/Main.hs:
```diff
@@ -98,23 +93,18 @@ data UniswapContracts =
+        UniswapStart   -> SomeBuiltin Uniswap.ownerEndpoint
+        Init           -> SomeBuiltin US.setupTokens
 
 handlers :: SimulatorEffectHandlers (Builtin UniswapContracts)
```
is this still needed?

koslambrou replied:
It's used in the `main` of the same file with `Simulator.runSimulationWith handlers $ do ...`

j-mueller commented on plutus-pab/src/Plutus/PAB/App.hs:
```diff
@@ -71,7 +73,7 @@ import           Servant.Client                                 (ClientEnv, mkCl
 ------------------------------------------------------------
 
-data AppEnv =
+data AppEnv a =
```
Comment explaining what `a` is

koslambrou replied:
Is the comment adequate?

j-mueller commented on plutus-pab/src/Plutus/PAB/Core/ContractInstance.hs:
```diff
@@ -82,16 +119,10 @@ activateContractSTM ::
-    logInfo @(ContractInstanceMsg t) $ ActivatedContractInstance caID caWallet activeContractInstanceId
-    pure activeContractInstanceId
+activateContractSTM runAppBackend a@ContractActivationArgs{caID} = do
+  activeContractInstanceId <- ContractInstanceId <$> uuidNextRandom
+  initialState <- Contract.initialState @t activeContractInstanceId caID
```
this could be extracted into a separate function that returns a `ContractInstanceState t`

j-mueller commented on plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs:
```diff
@@ -104,8 +118,17 @@ handleContractStore = \case
-            fromMaybe (throwError $ ContractStateNotFound instanceId)
-                      (pure <$> (_contractInstanceState c >>= decodeText))
+            do
+              caID <- maybe (throwError $ AesonDecodingError "Couldn't this JSON decode as `a`" (_contractInstanceId c))
```
We could use [`typeRep`](https://hackage.haskell.org/package/base-4.10.1.0/docs/Data-Typeable.html#v:typeRep) to get the name of the type for a more user-friendly error message

j-mueller commented on plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs:
```diff
@@ -93,31 +100,58 @@ instance PABContract (Builtin a) where
+    getSchema :: a -> [FunctionSchema FormSchema]
+
+-- | Defined in order to prevent type errors like: "Couldn't match type 'effs'
+-- with 'effs1'".
```
👍

j-mueller commented on plutus-pab/src/Plutus/PAB/Effects/Contract/Builtin.hs:
```diff
@@ -93,31 +100,58 @@ instance PABContract (Builtin a) where
     type State (Builtin a) = SomeBuiltinState a
     serialisableState _ = getResponse
 
+class HasDefinitions a where
```
The typeclass needs some haddocks

koslambrou replied:
Added some haddocks. Adequate enough?

j-mueller commented on plutus-pab/src/Plutus/PAB/Run.hs:
```diff
@@ -0,0 +1,88 @@
+    )
+    => BuiltinHandler a
+    -> IO ()
+runWith userContractHandler = do
```
Needs haddocks

Something like "PAB entry point for a contract type `a`"

koslambrou replied:
Added some comments. Adequate enough?

j-mueller commented on plutus-pab/src/Plutus/PAB/Run/PSGenerator.hs:
```diff
@@ -47,6 +51,9 @@ import           Servant.PureScript                         (HasBridge, Settings
                                                              writeAPIModuleWithSettings)
 
+class HasPSTypes a where
+  psTypes :: Proxy a -> [SumType 'Haskell]
```
Haddocks, we could also add a default definition that returns the empty list (I'm assuming that some (most?) clients won't be written in Purescript so they don't need the PS bridge)

koslambrou replied:
Done. Added default implement. Haddocks adequate?

j-mueller reviewed (commented):
This turned into a rather large change (incl. @silky 's work on restoring the contract state) but it's looking quite good now!

merivale reviewed (changes requested):
PureScript stuff looks okay. There are a few follow-on changes that'll be required, but I can do those in a subsequent PR. Just don't delete my comments please. :stuck_out_tongue: 

I'm wondering what the best way would be to generate PS types for the builtin contracts used in particular cases - like the `Marlowe` data type in the marlowe-pab case (please rename this to `MarloweContract`). We might just need a separate PSGenerator for that, but I mention it here in case someone has a better idea...

koslambrou commented on marlowe-dashboard-client/plutus-pab.yaml:
```diff
@@ -1,25 +1,19 @@
-# also determines which contracts are available). In most cases we use the
-# default settings - the main thing to watch out for is that the
-# `scZeroSlotTime` here coincides with the frontend code's assumptions about
-# the time of slot 0 (in `web-common-marlowe/src/Marlowe/Slot.purs`).
```
My fault. I put back those comments.

koslambrou commented on marlowe-dashboard-client/plutus-pab.yaml:
```diff
@@ -37,28 +31,23 @@ nodeServerConfig:
-# this isn't an issue here. However, there is a split second between creating
-# a contract and the endpoints first becoming available - so having this set
-# ensures that the first attempt to call an endpoint (immediately after
-# creating the contract) won't fail.
```
My fault. I put back those comments.

koslambrou commented on marlowe-dashboard-client/src/Capability/Marlowe.purs:
```diff
@@ -104,10 +103,10 @@ instance monadMarloweAppM :: ManageMarlowe AppM where
             ajaxCompanionAppId <- case pabType of
-              Plain -> Contract.activateContract (plutusAppPath WalletCompanion) wallet
+              Plain -> Contract.activateContract WalletCompanion wallet
               WithMarloweContracts -> Contract.activateContract WalletCompanion wallet
```
In that case, I'll let you sort this in a subsequent PR. Thanks!

koslambrou commented on marlowe-dashboard-client/default.nix:
```diff
@@ -1,5 +1,23 @@
+    $(nix-build ../default.nix --quiet --no-build-output -A plutus-pab.server-setup-invoker)/bin/plutus-pab-setup psgenerator $generatedDir
+    $(nix-build ../default.nix --quiet --no-build-output -A plutus-pab.test-generator)/bin/plutus-pab-test-psgenerator $generatedDir
+    $(nix-build ../default.nix --quiet --no-build-output -A marlowe-dashboard.marlowe-invoker)/bin/marlowe-pab --config plutus-pab.yaml psapigenerator $generatedDir
+  '';
```
You're right, my fault. I changed `generated-purescript` by calling `psgenerator` and passing the config file for `psapigenerator`.

koslambrou replied:
What's the point of using the config file to specify the contracts that are being used? We specify the available definitions with the `HasDefinitions` typeclass. You had something else in mind?

merivale replied:
I was vaguely wondering if it could be used to generate PS types for the `a` in `Builtin a`, but that's probably not possible. Don't worry about it.

koslambrou commented on marlowe-dashboard-client/package.json:
```diff
@@ -10,7 +10,7 @@
+    "start": "npm install && marlowe-pab-generate-purs && npm run purs:compile && npm run webpack:server",
     "link-contracts": "nix-build ../default.nix -A marlowe-dashboard.contractsJSON -o contracts",
     "install-contracts": "nix-build ../default.nix -A marlowe-dashboard.install-marlowe-contracts",
     "setup-contracts": "npm run install-contracts && npm run link-contracts"
```
Removed them. Thanks!

koslambrou commented on marlowe-dashboard-client/src/API/Contract.purs:
```diff
@@ -47,15 +47,6 @@ class
   getAllContractInstances :: forall m. MonadError AjaxError m => MonadAff m => a -> m (Array (ContractInstanceClientState a))
   getContractDefinitions :: forall m. MonadError AjaxError m => MonadAff m => a -> m (Array (ContractSignatureResponse a))
 
-instance contractExeContractActivationId :: ContractActivationId ContractExe where
```
Thanks :) I'll let you sort it in a subsequent PR.

koslambrou commented on marlowe/pab/MarloweContracts.hs:
```diff
@@ -0,0 +1,65 @@
+import qualified Plutus.PAB.Simulator                as Simulator
+import           Text.Read                           (readMaybe)
+
+data Marlowe =
```
Done!

koslambrou commented on plutus-pab-client/src/View/Pretty.purs:
```diff
@@ -33,9 +32,6 @@ withHeading prefix content =
     , content
     ]
 
-instance prettyContractExe :: Pretty ContractExe where
-  pretty ((ContractExe { contractPath })) = text contractPath
```
Mmm.... It's not really possible to create an instance of Pretty for Builtin, since it doesn't really have a data constructor with the contract.

@j-mueller How do you think this can be done? Do we need to use `ContractDef` to make that happen?

j-mueller replied:
I don't think we need a `prettyBuiltin` - the `Builtin` type has no values, so there isn't anything to pretty print. On the Haskell side we use `Builtin` with a type family. For example, the log messages for a `Builtin a` usually have `ContractDef a` as the concrete type. For `MarloweContract` this is just `MarloweContract` - so we'd need a `prettyMarloweContract`, but the instance needs to be written for each `a` individually, we can't provide a general one as part of `plutus-pab-client`.

koslambrou commented:
@merivale I can't see to be able to compile the module `Language.Marlowe.Client` in `marlowe` because it times out. Memory usage goes incredibly high. How you ever gotten that?

kwxm commented:
> @merivale I can't seem to be able to compile the module `Language.Marlowe.Client` in `marlowe` because it times out. Memory usage goes incredibly high. How you ever gotten that?

I concur.  I tried to compile Marlowe on our benchmarking machine (which has 8 cores and 32G of memory) and it took over 9 minutes to compile this file, with memory usage peaking over 16G (in the latest version of `master`).  I tried on my laptop as well: it quickly locked it up completely and I had to reboot!

merivale commented:
I haven't experienced that myself, but it might be the problem @palas mentioned he was having the other day. He thought it might just be his branch, but this makes me wonder whether something unfortunate has made it's way onto master... You should speak to him and @nau about this.

koslambrou commented:
@j-mueller A few changes since your last review. I delete the module `Plutus.PAB.ParseStringifiedJSON` and removed all log ADTs in `Plutus.PAB.Monitoring.PABLogMsg` which weren't used in the pab.

I also added the CLI command `available` which basically outputs the result of `Contracts.getDefinitions`. It basically replaces the old `installed` command which lists installed contracts.

j-mueller commented:
@koslambrou :

> added the CLI command `available`

Good idea. I'm happy to merge it now, if we can just get the build to finish!

@merivale : I looked into the marlowe client compilation issue a little bit [here](https://github.com/input-output-hk/plutus/pull/3593) - seems like it is caused by the compilation of the on-chain code bits, which we can't do much about.

koslambrou commented:
@j-mueller All that's left before merging is for @silky to apply your review for his part of the commit.

silky commented:
@koslambrou cool I'll do those now
